File: Trees/0988. Smallest String Starting From Leaf/0988.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ""
            
        self.res=[]
        def dfs(node,currString):
            if not node.left and not node.right:
                currString+=chr(node.val+ord('a'))
                self.res.append(currString[::-1])
                return
            currString+=chr(node.val+ord('a'))
            if node.left:
                dfs(node.left,currString)
            if node.right:
                dfs(node.right,currString)
        
        dfs(root,"")
        return sorted(self.res)[0]

File: Trees/0988. Smallest String Starting From Leaf/0988.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#pragma GCC optimize('03','unroll-loops')

class Solution {
public:
    vector<string>res;
    void dfs(TreeNode* node, string currString) {
        if(!node->left && !node->right){
            currString+=char(node->val+'a');
            reverse(currString.begin(),currString.end());
            res.push_back(currString);
            return;    
        }
        currString+=char(node->val+'a');
        if(node->left){
            dfs(node->left, currString);
        }
        if(node->right){
            dfs(node->right,currString);
        }
    }

    string smallestFromLeaf(TreeNode* root) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        if(!root){
            return "";
        }
        dfs(root,"");
        sort(res.begin(), res.end()); 
        return res[0];
    }
};